### PR TITLE
Update manager.inc to allow mount of labeled filesystems

### PR DESCRIPTION
**Pull request by burnbasket**
1) Reordered checks by specicifity
is_devicefile will catch anything in /dev and its subdirectories, so I think it should be done last. (Sidenote: Maybe you wanted ^\/dev\/[^\/]+$ in functions.inc?)
2) Label will never match the full /dev/disk/by-label path, so I propose to match only the string after the last slash. (This might be beneficial for the uuid part, too, but I didn't make this change here.)
3) The final else statement only handled the "label" case (and didn't in case the label wasn't on its own ;-) making it unnecessary. The proposed change should handle isolated labels as well as labels with path, because labels can't contain slashes afaik.
4) I didn't use else-if constructs, because the label case has a sub condition with a regex match and I didn't know how to do that as an && in the first condition. To keep things consistent used !found in all conditions except obviously the first.

I've never coded in PHP before, so please forgive me if it looks clumsy.

- [ ] References issue
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by votdev** [comment]
Can you add some example device names in the code to make it more clear what happens here.

Please also add a description why this PR is necessary, currently i do not understand the intention.

Please add the signed-off-by line (`Signed-off-by: Frank Mustermann <frank.mustermann@xxx.yyy>`).

**Comment by burnbasket** [comment]
Mount button threw exception with this device:
/dev/disk/by-label/openmediavaultData
It ended up being handled in the is_devicefile clause, because line 313 in functions.inc matches everything in /dev and its subdirectories.
preg_match('/^\/dev\/.+$/i', $var)

**Comment by burnbasket** [comment]
Both is_fs_uuid and is_devicefile_by in functions.inc are more specific.
I used is_devicefile_by because it was already there.

**Comment by votdev** [comment]
How can I reproduce this? I never saw that behavior and the code is really old there.

**Comment by burnbasket** [comment]
The only thing I did was create a volume with label, in this case
mkfs.btrfs --label openmediavaultData --metadata raid10 --data raid10 /dev/disk/by-id/ata-TOSHIBA_DT01ACA300_<XXX> /dev/disk/by-id/ata-SAMSUNG_HD204UI_<XXX> /dev/disk/by-id/ata-SAMSUNG_HD204UI_<XXX> /dev/disk/by-id/ata-SAMSUNG_HD103SI_<XXX>

**Comment by burnbasket** [comment]
But the fs really doesn't matter.

**Comment by burnbasket** [comment]
`function is_devicefile($var) {
	if (FALSE === is_string($var))
		return FALSE;
	return preg_match('/^\/dev\/.+$/i', $var) ? TRUE : FALSE;
}`
will alway match any string starting with /dev/

**Comment by burnbasket** [comment]
This could be changed by using this instead:

> /^\/dev\/[^\/]+$/

**Comment by votdev** [comment]
> Mount button threw exception with this device:
> /dev/disk/by-label/openmediavaultData
> It ended up being handled in the is_devicefile clause, because line 313 in functions.inc matches everything in /dev and its subdirectories.
> preg_match('/^/dev/.+$/i', $var)

That it the intention of this function to check if it is a device path, and that also includes /dev/disk/by... paths. Changing this will break everything in OMV, so this would be a really bad idea.

**Comment by burnbasket** [comment]
OK, but in this case, the label case can never be reached.

**Comment by votdev** [comment]
Please post the output of blkid and the dir tree of the device, e.g. /dev/... I can not merge this PR before I fully can reproduce this, otherwise it might break other installations.

**Comment by burnbasket** [comment]
I can see that.
`blkid /dev/disk/by-label/openmediavaultData -o full
/dev/disk/by-label/openmediavaultData: LABEL="openmediavaultData" UUID="6f22c380-dc25-424c-bb08-5495048b58dd" UUID_SUB="b8cd5870-c357-4f76-bb87-13a50625786e" TYPE="btrfs"`

**Comment by burnbasket** [comment]
label is a special case of dev imo

**Comment by votdev** [comment]
I need to check the code, but it works for me in the past and there were no bug reports related this till now. I assume it has something to do with BTRFS. I think a device by path link should be resolved to a canonical path when using realpath, so a previous code path should be executed in this case. But as said, I need to check that.
Changing code here is really dangerous, so we need to take care not to break everything.

**Comment by burnbasket** [comment]
functions.inc condition
`function is_devicefile_by($var) {
	if (FALSE === is_string($var))
		return FALSE;
	return preg_match('/^\/dev\/disk\/by-\S+\/.+$/i', $var) ? TRUE : FALSE;
}`
is more specific than
`function is_devicefile($var) {
	if (FALSE === is_string($var))
		return FALSE;
	return preg_match('/^\/dev\/.+$/i', $var) ? TRUE : FALSE;
}`

**Comment by burnbasket** [comment]
I really get that.

**Comment by burnbasket** [comment]
I was a little surprised at first, too.

**Comment by burnbasket** [comment]
I thought GUI mounts of dev sub directories looked like they would always fail for devices only specified by their label.

**Comment by burnbasket** [comment]
A workaround would be to use one of the constituent drives instead of the label, but that didn't jive with me.

**Comment by burnbasket** [comment]
`lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0 167.7G  0 disk 
├─sda1   8:1    0   512M  0 part /boot/efi
├─sda2   8:2    0 159.5G  0 part /
└─sda3   8:3    0   7.7G  0 part [SWAP]
sdb      8:16   1   1.8T  0 disk 
sdc      8:32   1   1.8T  0 disk 
sdd      8:48   1 931.5G  0 disk 
sde      8:64   1   2.7T  0 disk`

**Comment by burnbasket** [comment]
`tree /dev/disk
/dev/disk
├── by-id
│   ├── ata-INTEL_SSDSC2CT180A3____ -> ../../sda
│   ├── ata-INTEL_SSDSC2CT180A3____-part1 -> ../../sda1
│   ├── ata-INTEL_SSDSC2CT180A3____-part2 -> ../../sda2
│   ├── ata-INTEL_SSDSC2CT180A3____-part3 -> ../../sda3
│   ├── ata-SAMSUNG_HD103SI____ -> ../../sdd
│   ├── ata-SAMSUNG_HD204UI____ -> ../../sdc
│   ├── ata-SAMSUNG_HD204UI____ -> ../../sdb
│   ├── ata-TOSHIBA_DT01ACA300____ -> ../../sde
│   ├── wwn-0x5000039fe6ed496c -> ../../sde
│   ├── wwn-0x5001517bb29361b4 -> ../../sda
│   ├── wwn-0x5001517bb29361b4-part1 -> ../../sda1
│   ├── wwn-0x5001517bb29361b4-part2 -> ../../sda2
│   ├── wwn-0x5001517bb29361b4-part3 -> ../../sda3
│   ├── wwn-0x50024e90013839d6 -> ../../sdd
│   ├── wwn-0x50024e90040f7928 -> ../../sdc
│   └── wwn-0x50024e90040f7935 -> ../../sdb
├── by-label
│   └── openmediavaultData -> ../../sde
├── by-partuuid
│   ├── 28529ae3-3cac-4860-85c0-c744453f83c6 -> ../../sda1
│   ├── 5a55c57b-283b-4cf0-95b2-11228f578b03 -> ../../sda2
│   └── 630ee4a7-95fb-46c1-b82a-212613e5bba5 -> ../../sda3
├── by-path
│   ├── pci-0000:00:13.0-ata-1 -> ../../sda
│   ├── pci-0000:00:13.0-ata-1-part1 -> ../../sda1
│   ├── pci-0000:00:13.0-ata-1-part2 -> ../../sda2
│   ├── pci-0000:00:13.0-ata-1-part3 -> ../../sda3
│   ├── pci-0000:01:00.0-ata-1 -> ../../sdb
│   ├── pci-0000:01:00.0-ata-3 -> ../../sdc
│   └── pci-0000:01:00.0-ata-4 -> ../../sde
└── by-uuid
    ├── 25cf0f95-165a-4551-a6c5-38fa3da9a4a2 -> ../../sda3
    ├── 470cd4dc-d89a-4b0e-a56f-d5e7b5b2abb4 -> ../../sda2
    ├── 607C-4827 -> ../../sda1
    └── 6f22c380-dc25-424c-bb08-5495048b58dd -> ../../sde`

**Comment by burnbasket** [comment]
On the topic of canonical file paths, why not make the (unchanging) device ids or the (editable) labels the main anchor instead of the almost random /dev/sd* device order?

I realize you're on the same page as the kernel there, and you can't argue with the kernel, so these boots are made for walking.

Devices keep changin' when they ought to be samin' and some of these days these boots are walking all over us.

:-)

Happy Holidays!

**Comment by votdev** [comment]
> On the topic of canonical file paths, why not make the (unchanging) device ids or wwns or the (editable) labels the main anchor instead of the almost random /dev/sd* device order?
> 
> I realize you're on the same page as the kernel there, and you can't argue with the kernel, so these boots are made for walking.
> 
> Devices keep changin' when they ought to be samin' and some of these days these boots are walking all over us.
> 
> :-)
> 
> Happy Holidays!

Because of that OMV prefers predictable device names for storing in the database and for configuration files, but canonical device names are the best to work at runtime because they are simple.

**Comment by burnbasket** [comment]
I did some quickie tests with unmodded an modified filesystem/manager.inc:
TL;DR: For the tested scanarios, original and modded version behaved precisely the same.

Scenarios:
Ext2 filesystem on partition without label
Ext2 filesystem on partition with label
Ext2 filesystem on unpartitioned device without label
Ext2 filesystem on unpartitioned device with label

All these scenarios where performed with the original and with the modified version of filesystem/manager.inc, with poweroff and new boot between changes of this file.

For each scenario, the disk on which the test was performed was initialised with 8MB of zeros, then partitioned or not as required by the scenario and a new filesystem created on the cli. Where required, the filesystem was then labeled.

More details in the next post.

**Comment by burnbasket** [comment]
Here is the full version of the tests.

As mentioned in the last post, for each scenario, the disk on which the test was performed was initialised with 8MB of zeros, then partitioned or not as required by the scenario and a new filesystem created on the cli. Where required, the filesystem was then labeled.
##########
##### Test object: "Storage/File Systems" GUI item
## scope: gui component test
## Modified and unmodified artifact "filesystem/manager.inc"
## before test cleanup with cli command
~# for i in sdb sdc sdd sde; do dd if=/dev/zero of=/dev/$i bs=8M count=1; done

## original block device structure:
~# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0 167.7G  0 disk
├─sda1   8:1    0   512M  0 part /boot/efi
├─sda2   8:2    0 159.5G  0 part /
└─sda3   8:3    0   7.7G  0 part [SWAP]
sdb      8:16   1   1.8T  0 disk
sdc      8:32   1   1.8T  0 disk
sdd      8:48   1 931.5G  0 disk
sde      8:64   1   2.7T  0 disk

## Switched from unmodified to modified version by inserting new file, full shutdown and new boot.
## Presence of modified and unmodified filesystem/manager.inc verified by checking content of
sftp://root@10.1.1.11/usr/share/php/openmediavault/system/filesystem/backend/manager.inc
###########
## Original status:
procedure: on cli run lsblk
expected: only system fs visible
result:   only system fs visible

~# lsblk
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda      8:0    0 167.7G  0 disk
├─sda1   8:1    0   512M  0 part /boot/efi
├─sda2   8:2    0 159.5G  0 part /
└─sda3   8:3    0   7.7G  0 part [SWAP]
sdb      8:16   1   1.8T  0 disk
sdc      8:32   1   1.8T  0 disk
sdd      8:48   1 931.5G  0 disk
└─sdd1   8:49   1     2G  0 part
sde      8:64   1   2.7T  0 disk

############
## scenario: create fs on first partition of device, mount, umount and delete from GUI
############
## Prepare tests:
procedure: on cli create one partition on /dev/sdd, format it as ext2 and label it as "omvTestLabel00"
~# cfdisk /dev/sdd
  Select label type: gpt
  new /dev/sdd1 2G partitiontype "Linux filesystem"

~# mkfs.ext2 /dev/sdd1
mke2fs 1.45.3 (14-Jul-2019)
Creating filesystem with 524288 4k blocks and 131072 inodes
Filesystem UUID: 131a6f27-1180-4b19-91a6-a7ac1479762a
Superblock backups stored on blocks:
        32768, 98304, 163840, 229376, 294912

Allocating group tables: done
Writing inode tables: done
Writing superblocks and filesystem accounting information: done

~# e2label /dev/sdd1 omvTestLabel00


## Mount test
procedure: select tested filesystem, click buttons mount and apply
expected: mount and apply buttons in gui work
result:   mount and apply buttons in gui work
## Verify mount test:
procedure: on cli run mount without arguments
expected: /dev/sdd1 mounted
result:   /dev/sdd1 mounted
## Umount test:
procedure: select tested filesystem, click buttons unmount and apply
expected: unmount and apply buttons in gui work
result:   unmount and apply buttons in gui work
## Verify umount test:
procedure: on cli run mount without arguments
expected: /dev/sdd1 not mounted
result:   /dev/sdd1 not mounted
## Delete test:
procedure: select tested filesystem, click buttons delete and apply
expected: delete and apply buttons in gui work
result:   delete and apply buttons in gui work
## Verify delete test:
procedure: on cli run e2fsck /dev/sdd1 (no other arguments)
expected: /dev/sdd1 without filesystem
result:   /dev/sdd1 repaired by command from backup superblocks
comment:  result was not expected but is ok because of redundant fs data


############
## scenario: create fs on unpartitioned device, mount, umount and delete from GUI
############
##### both modded and original version give (imo unrelated) error on "apply":
##### modded with and without label: error works, error on apply, apply goes into loop, umount works, apply on umount works, delete works, e2fsck repairs

## Prepare tests:
procedure: on cli format raw /dev/sdd as ext2 and label it as "omvTestLabel00"

~# mkfs.ext2 /dev/sdd
mke2fs 1.45.3 (14-Jul-2019)
Creating filesystem with 244190646 4k blocks and 61054976 inodes
Filesystem UUID: d311bad2-be7f-4559-8d74-6ba3e828d3fd
Superblock backups stored on blocks:
        32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208,
        4096000, 7962624, 11239424, 20480000, 23887872, 71663616, 78675968,
        102400000, 214990848

Allocating group tables: done
Writing inode tables: done
Writing superblocks and filesystem accounting information: done

~# e2label /dev/sdd omvTestLabel00

## Mount test
procedure: select tested filesystem, click buttons mount and apply
expected: mount and apply buttons in gui work
result:   mount buttons in gui works, apply button produces error
## Verify mount test:
procedure: on cli run mount without arguments
expected: /dev/sdd mounted
result:   /dev/sdd mounted
## Umount test:
procedure: select tested filesystem, click buttons unmount and apply
expected: umount and apply buttons in gui work
result:   umount buttons in gui works, apply button produces error
## Verify umount test:
procedure: on cli run mount without arguments
expected: /dev/sdd not mounted
result:   /dev/sdd not mounted
## Delete test:
procedure: select tested filesystem, click buttons delete and apply
expected: delete and apply buttons in gui work
result:   delete and apply buttons in gui work
## Verify delete test:
procedure: on cli run e2fsck /dev/sdd1 (no other arguments)
expected: /dev/sdd without filesystem
result:   /dev/sdd repaired by command from backup superblocks
comment:  result was not expected but is ok because of redundant fs data

##### Remarks:
1) Ext2 was used for its similarity to Ext4 and to keep the testing times managable.
2) In all testcases the mount and umount operations the changes were active and visible on the cli without operating the "Apply" button.
3) The errors below appear to be unrelated to the test topic.
############# Sample errors #############

Error #0:
OMV\ExecException: Failed to execute command 'export PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin; export LANG=C.UTF-8; omv-salt deploy run collectd 2>&1' with exit code '1': debian:
    Data failed to compile:
----------
    Rendering SLS 'base:omv.deploy.collectd.plugins.disk' failed: Jinja error: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 394, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python3/dist-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python3/dist-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 51, in top-level template code
TypeError: 'NoneType' object is not subscriptable

; line 51

---
[...]
  #     "fsname": "008530ff-a134-4264-898d-9ce30eeab927",
  # }
  {% if salt['mount.is_mounted'](mountpoint.dir) %}
    {% set disk = salt['omv_utils.get_fs_parent_device_file'](mountpoint.fsname) %}
    # Extract the device name from '/dev/xxx'.
    {% set _ = disks.append(disk[5:]) %}    <======================
  {% endif %}
{% endfor %}
# Append the root filesystem.
{% set root_fs = salt['omv_utils.get_root_filesystem']() %}
{% set disk = salt['omv_utils.get_fs_parent_device_file'](root_fs) %}
[...]
--- in /usr/share/php/openmediavault/system/process.inc:182
Stack trace:
#0 /usr/share/php/openmediavault/engine/module/serviceabstract.inc(60): OMV\System\Process->execute()
#1 /usr/share/openmediavault/engined/rpc/config.inc(167): OMV\Engine\Module\ServiceAbstract->deploy()
#2 [internal function]: Engined\Rpc\Config->applyChanges(Array, Array)
#3 /usr/share/php/openmediavault/rpc/serviceabstract.inc(123): call_user_func_array(Array, Array)
#4 /usr/share/php/openmediavault/rpc/serviceabstract.inc(149): OMV\Rpc\ServiceAbstract->callMethod('applyChanges', Array, Array)
#5 /usr/share/php/openmediavault/rpc/serviceabstract.inc(588): OMV\Rpc\ServiceAbstract->OMV\Rpc\{closure}('/tmp/bgstatuswK...', '/tmp/bgoutputVE...')
#6 /usr/share/php/openmediavault/rpc/serviceabstract.inc(159): OMV\Rpc\ServiceAbstract->execBgProc(Object(Closure))
#7 /usr/share/openmediavault/engined/rpc/config.inc(189): OMV\Rpc\ServiceAbstract->callMethodBg('applyChanges', Array, Array)
#8 [internal function]: Engined\Rpc\Config->applyChangesBg(Array, Array)
#9 /usr/share/php/openmediavault/rpc/serviceabstract.inc(123): call_user_func_array(Array, Array)
#10 /usr/share/php/openmediavault/rpc/rpc.inc(86): OMV\Rpc\ServiceAbstract->callMethod('applyChangesBg', Array, Array)
#11 /usr/sbin/omv-engined(537): OMV\Rpc\Rpc::call('Config', 'applyChangesBg', Array, Array, 1)
#12 {main}







mkfs.ext2 /dev/sdd
mke2fs 1.45.3 (14-Jul-2019)
Creating filesystem with 244190646 4k blocks and 61054976 inodes
Filesystem UUID: 796af92b-eaee-4233-8021-372af4c572a3
Superblock backups stored on blocks:
        32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208,
        4096000, 7962624, 11239424, 20480000, 23887872, 71663616, 78675968,
        102400000, 214990848

Allocating group tables: done
Writing inode tables: done
Writing superblocks and filesystem accounting information: done

root@openmediavault:~# e2label /dev/sdd omvTestLabel00




An error has occured
Error #0:
OMV\ExecException: Failed to execute command 'export PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin; export LANG=C.UTF-8; omv-salt deploy run collectd 2>&1' with exit code '1': debian:
    Data failed to compile:
----------
    Rendering SLS 'base:omv.deploy.collectd.plugins.disk' failed: Jinja error: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/templates.py", line 394, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python3/dist-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python3/dist-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/lib/python3/dist-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 51, in top-level template code
TypeError: 'NoneType' object is not subscriptable

; line 51

---
[...]
  #     "fsname": "008530ff-a134-4264-898d-9ce30eeab927",
  # }
  {% if salt['mount.is_mounted'](mountpoint.dir) %}
    {% set disk = salt['omv_utils.get_fs_parent_device_file'](mountpoint.fsname) %}
    # Extract the device name from '/dev/xxx'.
    {% set _ = disks.append(disk[5:]) %}    <======================
  {% endif %}
{% endfor %}
# Append the root filesystem.
{% set root_fs = salt['omv_utils.get_root_filesystem']() %}
{% set disk = salt['omv_utils.get_fs_parent_device_file'](root_fs) %}
[...]
--- in /usr/share/php/openmediavault/system/process.inc:182
Stack trace:
#0 /usr/share/php/openmediavault/engine/module/serviceabstract.inc(60): OMV\System\Process->execute()
#1 /usr/share/openmediavault/engined/rpc/config.inc(167): OMV\Engine\Module\ServiceAbstract->deploy()
#2 [internal function]: Engined\Rpc\Config->applyChanges(Array, Array)
#3 /usr/share/php/openmediavault/rpc/serviceabstract.inc(123): call_user_func_array(Array, Array)
#4 /usr/share/php/openmediavault/rpc/serviceabstract.inc(149): OMV\Rpc\ServiceAbstract->callMethod('applyChanges', Array, Array)
#5 /usr/share/php/openmediavault/rpc/serviceabstract.inc(588): OMV\Rpc\ServiceAbstract->OMV\Rpc\{closure}('/tmp/bgstatusjW...', '/tmp/bgoutputn1...')
#6 /usr/share/php/openmediavault/rpc/serviceabstract.inc(159): OMV\Rpc\ServiceAbstract->execBgProc(Object(Closure))
#7 /usr/share/openmediavault/engined/rpc/config.inc(189): OMV\Rpc\ServiceAbstract->callMethodBg('applyChanges', Array, Array)
#8 [internal function]: Engined\Rpc\Config->applyChangesBg(Array, Array)
#9 /usr/share/php/openmediavault/rpc/serviceabstract.inc(123): call_user_func_array(Array, Array)
#10 /usr/share/php/openmediavault/rpc/rpc.inc(86): OMV\Rpc\ServiceAbstract->callMethod('applyChangesBg', Array, Array)
#11 /usr/sbin/omv-engined(537): OMV\Rpc\Rpc::call('Config', 'applyChangesBg', Array, Array, 1)
#12 {main}

**Comment by burnbasket** [comment]
Full version of the tests attached.
[omvTest20191224T004230.txt](https://github.com/openmediavault/openmediavault/files/3996618/omvTest20191224T004230.txt)

**Comment by votdev** [comment]
You still have not described the intention of the PR. Could you please do that? I'm now more irritated when i look into your attached file. I can see a Salt stack trace? Are you trying to fix that? If yes, why do you think this can be fixed in the PHP backend code that is not related to the Python code used by Salt?

**Comment by votdev** [comment]
> `tree /dev/disk /dev/disk ├── by-id │   ├── ata-INTEL_SSDSC2CT180A3____ -> ../../sda │   ├── ata-INTEL_SSDSC2CT180A3____-part1 -> ../../sda1 │   ├── ata-INTEL_SSDSC2CT180A3____-part2 -> ../../sda2 │   ├── ata-INTEL_SSDSC2CT180A3____-part3 -> ../../sda3 │   ├── ata-SAMSUNG_HD103SI____ -> ../../sdd │   ├── ata-SAMSUNG_HD204UI____ -> ../../sdc │   ├── ata-SAMSUNG_HD204UI____ -> ../../sdb │   ├── ata-TOSHIBA_DT01ACA300____ -> ../../sde │   ├── wwn-0x5000039fe6ed496c -> ../../sde │   ├── wwn-0x5001517bb29361b4 -> ../../sda │   ├── wwn-0x5001517bb29361b4-part1 -> ../../sda1 │   ├── wwn-0x5001517bb29361b4-part2 -> ../../sda2 │   ├── wwn-0x5001517bb29361b4-part3 -> ../../sda3 │   ├── wwn-0x50024e90013839d6 -> ../../sdd │   ├── wwn-0x50024e90040f7928 -> ../../sdc │   └── wwn-0x50024e90040f7935 -> ../../sdb ├── by-label │   └── openmediavaultData -> ../../sde ├── by-partuuid │   ├── 28529ae3-3cac-4860-85c0-c744453f83c6 -> ../../sda1 │   ├── 5a55c57b-283b-4cf0-95b2-11228f578b03 -> ../../sda2 │   └── 630ee4a7-95fb-46c1-b82a-212613e5bba5 -> ../../sda3 ├── by-path │   ├── pci-0000:00:13.0-ata-1 -> ../../sda │   ├── pci-0000:00:13.0-ata-1-part1 -> ../../sda1 │   ├── pci-0000:00:13.0-ata-1-part2 -> ../../sda2 │   ├── pci-0000:00:13.0-ata-1-part3 -> ../../sda3 │   ├── pci-0000:01:00.0-ata-1 -> ../../sdb │   ├── pci-0000:01:00.0-ata-3 -> ../../sdc │   └── pci-0000:01:00.0-ata-4 -> ../../sde └── by-uuid ├── 25cf0f95-165a-4551-a6c5-38fa3da9a4a2 -> ../../sda3 ├── 470cd4dc-d89a-4b0e-a56f-d5e7b5b2abb4 -> ../../sda2 ├── 607C-4827 -> ../../sda1 └── 6f22c380-dc25-424c-bb08-5495048b58dd -> ../../sde`

I'm sorry, i can really get no information out of this.

Please explain the problem and how to reproduce it. What i know out of the current information is:

* You can not mount a BTRFS filesystem which has a label.

I need:
* The error reported by the RPC in the UI error dialog
* The `blkid` output (without any command parameter)
* If you know the device file, then i want to have
```
# ls -alh /dev/disk/by-xxx/foo
# ls -alh /dev/disk/by-xxx/
```

**Comment by votdev** [comment]
Can not reproduce this behavior.

Peek 2019-12-24 01-17

**Comment by votdev** [comment]
I've checked the original code again and IMO you didn't have understood it correct. The `getBackendById` method can be called with a filesystem label as argument ([see method docu](https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/php/openmediavault/system/filesystem/backend/manager.inc#L111)). That's why the last else path is checking for the label. Extracting the label from a device path is dangerous and prohibited if you do not unescape it, otherwise you will not get the result you expect. If you call the method with a device file like `/dev/disk/by-label/omvTestLabel00` then the line `else if (realpath($id) == realpath($enumv['devicefile']))` ([see line 138](https://github.com/openmediavault/openmediavault/blob/master/deb/openmediavault/usr/share/php/openmediavault/system/filesystem/backend/manager.inc#L138)) should ALWAYS match if the filesystem is detected by `$enums = $this->enumerate();` because when you compare the canonical paths of both device files they should always match, either the enum method return a different device file for the filesystem.
```
root@omv5box:~# ls -alh /dev/disk/by-label/omvTestLabel00
lrwxrwxrwx 1 root root 9 Dec 24 00:16 /dev/disk/by-label/omvTestLabel00 -> ../../sda
```
IMO the current implementation is correct and the problem comes from elsewhere .

**Comment by votdev** [comment]
I will close this PR because i do not see a problem here. We should open a bug report instead, describing the problem, providing all available information and see how to fix the issue.

**Comment by burnbasket** [comment]
Thanks, I'll do that.

**Comment by burnbasket** [comment]
00_initializeDisksAndCreateFilesystem
01_filesystemDetectedInOmv
02_mountErrorInOmv
03_mountErrorDetailsInOmv
04_lsblkOnThisMachine
05_diskTreeOnThisMachine
06_mountOutputAfterGuiMountError

**Comment by votdev** [comment]
Please post the output of `# blkid`.

**Comment by burnbasket** [comment]
root@openmediavault:~# blkid
/dev/sda1: UUID="710D-6276" TYPE="vfat" PARTUUID="34b64675-37c5-4609-bf7b-3a7583581c2c"
/dev/sda2: UUID="540e1951-019c-4745-88dd-f91788f2521d" TYPE="ext4" PARTUUID="0ed5021c-d409-4f2b-b83e-9root@openmediavault:~# blkid
/dev/sda1: UUID="710D-6276" TYPE="vfat" PARTUUID="34b64675-37c5-4609-bf7b-3a7583581c2c"
/dev/sda2: UUID="540e1951-019c-4745-88dd-f91788f2521d" TYPE="ext4" PARTUUID="0ed5021c-d409-4f2b-b83e-92c78c76a45a"
/dev/sda3: UUID="b0458068-f10f-4be9-ac01-2581d5db6990" TYPE="swap" PARTUUID="8e762cad-31b1-4c3f-8503-29bcd80cf353"
/dev/sdb: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="f401beba-fd9b-4ef5-ad68-c2b37419c40b" TYPE="btrfs"
/dev/sdc: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="12e44150-85d5-4c9f-9b56-0a012a8755f1" TYPE="btrfs"
/dev/sdd: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="71810b69-f622-4c8f-b6fa-5247d2e98161" TYPE="btrfs"
/dev/sde: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="92c1c012-5130-448e-8c9b-63c86410b9b9" TYPE="btrfs"
Screenshot_20191224_151010
2c78c76a45a"
/dev/sda3: UUID="b0458068-f10f-4be9-ac01-2581d5db6990" TYPE="swap" PARTUUID="8e762cad-31b1-4c3f-8503-29bcd80cf353"
/dev/sdb: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="f401beba-fd9b-4ef5-ad68-c2b37419c40b" TYPE="btrfs"
/dev/sdc: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="12e44150-85d5-4c9f-9b56-0a012a8755f1" TYPE="btrfs"
/dev/sdd: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="71810b69-f622-4c8f-b6fa-5247d2e98161" TYPE="btrfs"
/dev/sde: LABEL="openmediavaultData" UUID="b1ce5d31-00c0-4542-a9b1-295c39618b03" UUID_SUB="92c1c012-5130-448e-8c9b-63c86410b9b9" TYPE="btrfs"

**Comment by votdev** [comment]
The real issue here is way more complicated than you tried to fix with your PR. BTRFS has a strange behaviour with canonical device files and what the kernel reports in the mount table.

**Comment by burnbasket** [comment]
Oh boy.
Well, for the time being my hack works on the old beast I was playing around with on my vacation, which is all I originally really intended.
I just thought maybe someone else might benefit if I upstreamed it and I was unaware of the development repository.

**Comment by burnbasket** [comment]
BTW: Thanks for the fast reaction!
In case this is hardware related (the machine I am running this on is made from spit, baling wire and stuff a dumpster diver could have pulled out of the trash, after all), I'm attaching the output of
lspci -vvv
[lspciVvv.txt](https://github.com/openmediavault/openmediavault/files/3998954/lspciVvv.txt)
